### PR TITLE
feat: add support for new cloud client test framework in google-cloud-spanner-executor

### PR DESCRIPTION
**Pull request by gyang-google**
This is the initial commit for client-wise executor of cloud Java Spanner client. This feature enables the ability to test cloud Spanner clients through an universal testing protos, defined by proto-google-cloud-spanner-executor-v1.

### CODE REVIEW

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java [hunk 1]
```diff
@@ -0,0 +1,3482 @@
...
+    Matcher matcher = DB_NAME.matcher(dbPath);
+    if (!matcher.matches()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Bad database path: " + dbPath);
+    }
+    return DatabaseId.of(matcher.group(1), matcher.group(2), matcher.group(3));
+  }
+
+  /** Build instance proto from cloud spanner instance. */
+  private com.google.spanner.admin.instance.v1.Instance instanceToProto(Instance instance) {
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3397]: Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceInfo.java#L320?

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java [hunk 2]
```diff
@@ -0,0 +1,3482 @@
...
+        break;
+      default:
+        throw new IllegalArgumentException("Unknown state:" + instance.getState());
+    }
+    instanceBuilder.setState(state);
+    return instanceBuilder.build();
+  }
+
+  /** Build instance proto from cloud spanner instance. */
+  private com.google.spanner.admin.instance.v1.InstanceConfig instanceConfigToProto(
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3428]: Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigInfo.java#L418?

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java [hunk 3]
```diff
@@ -0,0 +1,3482 @@
...
+  }
+
+  /** Get DatabaseId by parsing the given dbPath. */
+  private static DatabaseId getDatabaseId(String dbPath) {
+    Matcher matcher = DB_NAME.matcher(dbPath);
+    if (!matcher.matches()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, "Bad database path: " + dbPath);
+    }
+    return DatabaseId.of(matcher.group(1), matcher.group(2), matcher.group(3));
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3393]: Can we directly use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseId.java#L84?

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java [hunk 4]
```diff
@@ -0,0 +1,3482 @@
...
+          return com.google.cloud.spanner.Type.json();
+        }
+      default:
+        throw SpannerExceptionFactory.newSpannerException(
+            ErrorCode.INVALID_ARGUMENT, "Unsupported proto type: " + typeProto);
+    }
+  }
+
+  /** Convert a cloud Type to a Type proto. */
+  private static com.google.spanner.v1.Type cloudTypeToTypeProto(@NotNull Type cloudTypeProto) {
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3276]: Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java#L427?

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java [hunk 5]
```diff
@@ -0,0 +1,3482 @@
...
+    for (int i = 0; i < fieldTypes.size(); ++i) {
+      builder
+          .set(fieldTypes.get(i).getName())
+          .to(valueProtoToCloudValue(fieldTypes.get(i).getType(), fieldValues.get(i)));
+    }
+    return builder.build();
+  }
+
+  /** Convert a Type proto to a cloud Type. */
+  private static com.google.cloud.spanner.Type typeProtoToCloudType(
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3223]: Can we re-use https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/main/java/com/google/cloud/spanner/Type.java#L442?

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/WorkerProxy.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  private static final String OPTION_SPANNER_PORT = "spanner_port";
+  private static final String OPTION_PORT = "port";
+  private static final String OPTION_CERTIFICATE = "cert";
+  private static final String OPTION_KEY = "key";
+  private static final String OPTION_USE_PLAIN_TEXT_CHANNEL = "use_plain_text_channel";
+  private static final String OPTION_ENABLE_GRPC_FAULT_INJECTOR = "enable_grpc_fault_injector";
+
+  public static int spannerPort = 0;
+  public static int port = 0;
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/WorkerProxy.java:50]: `port` -> `proxyPort`?

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/WorkerProxy.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  public static String cert = "";
+  public static String key = "";
```
Comment (rajatbhatta) [inline google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/WorkerProxy.java:52]: `key` -> `serviceKeyFile`?

**Review by rajatbhatta — CHANGES_REQUESTED** [review]

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java (hunk 5 above)
Comment (gyang-google) [reply google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3223]: This API is not public.

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java (hunk 4 above)
Comment (gyang-google) [reply google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3276]: This API is not public.

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java (hunk 2 above)
Comment (gyang-google) [reply google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3428]: This API is not public.

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java (hunk 1 above)
Comment (gyang-google) [reply google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:3397]: This API is not public.

File: google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java [hunk 8]
```diff
@@ -0,0 +1,3482 @@
...
+      getTransactionForWrite().buffer(mutations);
+    }
+
+    /** Execute a batch of updates in a read-write transaction. */
+    public synchronized long[] executeBatchDml(@NotNull List<Statement> stmts)
+        throws SpannerException {
+      for (int i = 0; i < stmts.size(); i++) {
+        LOGGER.log(
+            Level.INFO, String.format("executeBatchDml [%d]: %s", i + 1, stmts.get(i).toString()));
+      }
```
[the remark this answers is not in the document]
Comment (gyang-google) [reply google-cloud-spanner-executor/src/main/java/com/google/cloud/executor/spanner/CloudClientExecutor.java:647]: I guess here we intend to print the index of the statement.

**Review by rajatbhatta — APPROVED** [review]
Lgtm. Thank you for working on this, @gyang-google !

**Pull request merged**